### PR TITLE
Deduplicate code

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
The markdown rendering component

### Description
This PR improves the code quality of `markdown-preview.tsx` by deduplicating some code blocks that are used to identify which function should replace a tag.

### Steps

- [x] added implementation
- [x] added / updated tests
- [x] added / updated documentation
- [x] extended changelog

### CODE REVIEW

File: src/components/editor/markdown-preview/markdown-preview.tsx [hunk 1]
```diff
@@ -12,23 +13,37 @@ import { replaceLegacySlideshareShortCode } from './regex-plugins/replace-legacy
...
 import { getVimeoReplacement } from './replace-components/vimeo/vimeo-frame'
 import { getYouTubeReplacement } from './replace-components/youtube/youtube-frame'
 
 export interface MarkdownPreviewProps {
   content: string
 }
 
+export type ComponentReplacer = (node: DomElement, counterMap: Map<string, number>) => (ReactElement | undefined);
+const allComponentReplacer: ComponentReplacer[] = [getYouTubeReplacement, getVimeoReplacement, getGistReplacement, getPDFReplacement]
+type Transformer2id2CounterMap = Map<ComponentReplacer, Map<string, number>>
```
Comment (DerMolly) [inline src/components/editor/markdown-preview/markdown-preview.tsx:30]:
```suggestion
type Transformer2ID2CounterMap = Map<ComponentReplacer, Map<string, number>>
```

File: src/components/editor/markdown-preview/markdown-preview.tsx [hunk 2]
```diff
@@ -12,23 +13,37 @@ import { replaceLegacySlideshareShortCode } from './regex-plugins/replace-legacy
...
+
+const findAReplacementComponent = (node: DomElement, transformer2id2CounterMap: Transformer2id2CounterMap) => {
+  return allComponentReplacer
+    .map((componentReplacer) => {
+      const id2CounterMap = transformer2id2CounterMap.get(componentReplacer) || new Map<string, number>()
```
Comment (DerMolly) [inline src/components/editor/markdown-preview/markdown-preview.tsx:35]:
```suggestion
      const ID2CounterMap = transformer2ID2CounterMap.get(componentReplacer) || new Map<string, number>()
```

File: src/components/editor/markdown-preview/markdown-preview.tsx [hunk 3]
```diff
@@ -46,28 +61,10 @@ const MarkdownPreview: React.FC<MarkdownPreviewProps> = ({ content }) => {
   }, [])
 
   const result: ReactElement[] = useMemo(() => {
-    const youtubeIdCounterMap = new Map<string, number>()
-    const vimeoIdCounterMap = new Map<string, number>()
-    const gistIdCounterMap = new Map<string, number>()
-
+    const transformer2id2CounterMap = new Map<ComponentReplacer, Map<string, number>>()
```
Comment (DerMolly) [inline src/components/editor/markdown-preview/markdown-preview.tsx:64]:
```suggestion
    const transformer2ID2CounterMap = new Map<ComponentReplacer, Map<string, number>>()
```

File: src/components/editor/markdown-preview/markdown-preview.tsx [hunk 4]
```diff
@@ -46,28 +61,10 @@ const MarkdownPreview: React.FC<MarkdownPreviewProps> = ({ content }) => {
...
-        return resultVimeo
-      }
-
-      const resultGist = getGistReplacement(node, gistIdCounterMap)
-      if (resultGist) {
-        return resultGist
-      }
-
-      return convertNodeToElement(node, index, transform)
+      return findAReplacementComponent(node, transformer2id2CounterMap) || convertNodeToElement(node, index, transform)
```
Comment (DerMolly) [inline src/components/editor/markdown-preview/markdown-preview.tsx:67]:
```suggestion
      return findAReplacementComponent(node, transformer2ID2CounterMap) || convertNodeToElement(node, index, transform)
```

**Comment by DerMolly** [comment]
Also there are still items unticked in the inital comment of the PR

**Comment by mrdrogdrog** [comment]
@DerMolly I renamed some of the vars, but with another pattern. Please close the conversations if it's okay for you.

File: src/components/editor/editor.tsx [hunk 5]
```diff
@@ -12,7 +12,22 @@ import { TaskBar } from './task-bar/task-bar'
...
+    'https://gist.github.com/schacon/1\n' +
+    '\n' +
+    '## YouTube\n' +
+    'https://www.youtube.com/watch?v=KgMpKsp23yY\n' +
+    '\n' +
+    '## Vimeo\n' +
+    'https://vimeo.com/23237102\n' +
+    '\n' +
+    '## PDF\n' +
+    '{%pdf https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf %}')
```
Comment (ErikMichelson) [inline src/components/editor/editor.tsx:30]: Can't we use template literals for this?

Comment (mrdrogdrog) [reply src/components/editor/editor.tsx]: done

**Pull request merged**